core/trino-spi/src/main/java/io/trino/spi/block/ByteArrayBlockEncoding.java

```diff
@@ -82,14 +86,54 @@ public ByteArrayBlock readBlock(BlockEncodingSerde blockEncodingSerde, SliceInpu
         int positionCount = sliceInput.readInt();
 
         byte[] valueIsNullPacked = retrieveNullBits(sliceInput, positionCount);
-        byte[] values = new byte[positionCount];
 
         if (valueIsNullPacked == null) {
-            sliceInput.readBytes(Slices.wrappedBuffer(values));
+            byte[] values = new byte[positionCount];
+            sliceInput.readBytes(values, 0, values.length);
             return new ByteArrayBlock(0, positionCount, null, values);
         }
-        boolean[] valueIsNull = decodeNullBits(valueIsNullPacked, positionCount);
+        else {
+            boolean[] valueIsNull = decodeNullBits(valueIsNullPacked, positionCount);
+            if (vectorizeNullCompressAndExpand) {
+                return expandBytesWithNullsVectorized(sliceInput, positionCount, valueIsNull);
+            }
+            else {
+                return expandBytesWithNullsScalar(sliceInput, positionCount, valueIsNullPacked, valueIsNull);
+            }
+        }
+    }
 
+    private static ByteArrayBlock expandBytesWithNullsVectorized(SliceInput sliceInput, int positionCount, boolean[] valueIsNull)
+    {
+        if (valueIsNull.length != positionCount) {
+            throw new IllegalArgumentException("valueIsNull length must match positionCount");
+        }
+        int nonNullPositionCount = sliceInput.readInt();
+        int nonNullIndex = positionCount - nonNullPositionCount;
+        byte[] values = new byte[positionCount];
+        sliceInput.readBytes(values, nonNullIndex, nonNullPositionCount);
+
+        int position = 0;
+        for (; position < nonNullIndex && nonNullIndex + BYTE_SPECIES.length() < values.length; position += BYTE_SPECIES.length()) {
+            ByteVector nonNullValues = ByteVector.fromArray(BYTE_SPECIES, values, nonNullIndex);
+            VectorMask<Byte> nonNullMask = BYTE_SPECIES.loadMask(valueIsNull, position).not();
+            nonNullIndex += nonNullMask.trueCount();
+            nonNullValues
+                    .expand(nonNullMask)
+                    .intoArray(values, position);
+        }
+        for (; position < nonNullIndex; position++) {
+            values[position] = valueIsNull[position] ? 0 : values[nonNullIndex++];
+        }
+        return new ByteArrayBlock(0, positionCount, valueIsNull, values);
+    }
+
+    private static ByteArrayBlock expandBytesWithNullsScalar(SliceInput sliceInput, int positionCount, byte[] valueIsNullPacked, boolean[] valueIsNull)
+    {
+        if (valueIsNull.length != positionCount) {
```

Comment:
We don't have such check in the existing impl, why do you want to add the check?